// threerest/js/hypermedia/halPaginateFlux.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _halFlux = require("./halFlux");

var _halFlux2 = _interopRequireDefault(_halFlux);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var NEXT_LINK_PROP = "nextLink";
var PREVIOUS_LINK_PROP = "previousLink";
var FIRST_LINK_PROP = "first";
var LAST_LINK_PROP = "last";

/**
 * Class for decorate data with HAL metadata.
 * @class
 */

var HalPaginateFlux = function (_HalFlux) {
  _inherits(HalPaginateFlux, _HalFlux);

  function HalPaginateFlux(data, paginationData) {
    _classCallCheck(this, HalPaginateFlux);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(HalPaginateFlux).call(this, data));

    _this.paginationData = paginationData;
    return _this;
  }

  _createClass(HalPaginateFlux, [{
    key: "updateLinks",
    value: function updateLinks(currentUrl, baseUrl) {
      _get(Object.getPrototypeOf(HalPaginateFlux.prototype), "updateLinks", this).call(this, currentUrl);
      var paginationMetadata = this.paginationData.getPageMetadata();
      this.nextLink = baseUrl + "?" + paginationMetadata.pageSizeKeyWord + "=" + paginationMetadata.pageSize + "&" + paginationMetadata.pageIdxKeyWord + "=" + paginationMetadata.nextIdx;
      this.previousLink = baseUrl + "?" + paginationMetadata.pageSizeKeyWord + "=" + paginationMetadata.pageSize + "&" + paginationMetadata.pageIdxKeyWord + "=" + paginationMetadata.prevIdx;
      this.firstLink = baseUrl + "?" + paginationMetadata.pageSizeKeyWord + "=" + paginationMetadata.pageSize + "&" + paginationMetadata.pageIdxKeyWord + "=0";
      this.lastLink = baseUrl + "?" + paginationMetadata.pageSizeKeyWord + "=" + paginationMetadata.pageSize + "&" + paginationMetadata.pageIdxKeyWord + "=" + paginationMetadata.lastPage;
    }
  }, {
    key: "nextLink",
    set: function set(value) {
      this._links[NEXT_LINK_PROP] = { href: value };
    }
  }, {
    key: "previousLink",
    set: function set(value) {
      this._links[PREVIOUS_LINK_PROP] = { href: value };
    }
  }, {
    key: "firstLink",
    set: function set(value) {
      this._links[FIRST_LINK_PROP] = { href: value };
    }
  }, {
    key: "lastLink",
    set: function set(value) {
      this._links[LAST_LINK_PROP] = { href: value };
    }
  }]);

  return HalPaginateFlux;
}(_halFlux2.default);

exports.default = HalPaginateFlux;
